cli/src/cli_util.rs

```diff
@@ -3779,8 +3780,15 @@ impl CliRunner {
         self
     }
 
-    pub fn add_start_hook(mut self, start_hook_fn: CliDispatchFn) -> Self {
-        self.start_hook_fns.push(start_hook_fn);
+    /// Add a hook that gets called when it's time to run the command. It is
+    /// the hook's responsibility to call the given inner dispatch function to
+    /// run the command.
+    pub fn add_dispatch_hook(mut self, dispatch_hook_fn: CliDispatchHookFn) -> Self {
```

Comment:
nit: `dispatch_hook_fn` doesn't have to be `Box<dyn ..>`.

And I think `.add_subcommand()` can be forwarded to `.add_dispatch_hook()` as an example. (This means the order of `.add_dispatch_hook()` and `.add_subcommand()` is important?)